File: crates/learning-data-access/src/live_gradebook.rs
//! Answer-free Instructor Gradebook projection of immutable Grading Results.

use async_trait::async_trait;
use question_model::CourseInstanceId;
use serde::Serialize;

use crate::{LiveAssessmentAttemptScore, SessionTokenHash, StoreError};

/// One Course-local answer-free progress row for an active Student.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseGradebookStudentWork {
    pub roster_id: String,
    /// Instructor-provided, exact-Course roster label.
    pub roster_name: String,
    pub assessment_id: question_model::AssessmentId,
    pub assessment_title: String,
    pub assessment_attempt_completion: Option<question_model::AssessmentAttemptCompletion>,
    /// Derived only from server time and durable submission evidence.
    pub expired_submitting: bool,
    /// Missing only until background submission records immutable outcomes.
    /// This is a Gradebook contribution pair: Bonus work can contribute earned
    /// points with zero possible, and extra credit can exceed possible points.
    pub score: Option<LiveAssessmentAttemptScore>,
}

/// The current Course Instructor's answer-free Gradebook evidence projection.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseGradebook {
    pub course_id: CourseInstanceId,
    pub student_work: Vec<CourseGradebookStudentWork>,
}

#[async_trait]
pub trait CourseGradebookStore: Send + Sync {
    /// Reads the exact Course Instructor's answer-free Student Work progress.
    /// Foreign and unavailable Courses remain concealed.
    async fn course_gradebook(
        &self,
        session_token_hash: SessionTokenHash,
        course: CourseInstanceId,
    ) -> Result<CourseGradebook, StoreError>;
}

File: src/api/live_gradebook.ts
// Browser-safe Instructor Gradebook evidence boundary.

import type { AssessmentId } from "../../generated/api/AssessmentId";
import type { AssessmentAttemptCompletion } from "../../generated/api/AssessmentAttemptCompletion";
import type { CourseInstanceId } from "../../generated/api/CourseInstanceId";

/** One answer-free progress aggregate for an active Student. */
export interface CourseGradebookStudentWork {
  readonly rosterId: string;
  readonly rosterName: string;
  readonly assessmentId: AssessmentId;
  readonly assessmentTitle: string;
  readonly assessmentAttemptCompletion: AssessmentAttemptCompletion | null;
  /** Derived from server time; no grading operation or queue state. */
  readonly expiredSubmitting: boolean;
  /**
   * Null while an expired Attempt is awaiting background submission. Gradebook
   * contributions can earn points with zero possible or exceed possible points.
   */
  readonly score: {
    readonly pointsEarned: number;
    readonly pointsPossible: number;
  } | null;
}

/** The complete browser projection for one current Instructor Course. */
export interface CourseGradebook {
  readonly courseId: CourseInstanceId;
  readonly studentWork: ReadonlyArray<CourseGradebookStudentWork>;
}

export type GradebookExportFormat = "csv" | "tsv";

/** Same-origin current-Instructor Gradebook read and point-export capability. */
export interface CourseGradebookClient {
  readonly getCourseGradebook: (courseInstanceId: CourseInstanceId) => Promise<CourseGradebook>;
  readonly downloadCourseGradebook: (
    courseInstanceId: CourseInstanceId,
    format: GradebookExportFormat,
  ) => Promise<Blob>;
}
